File: skybit/entities.py
import math
import random
import pyxel
from .config import (GRAVITY, FLAP_V, MAX_FALL, BIRD_X, BIRD_R,
                     GROUND_Y, PIPE_W, GAP_START, GAP_MIN,
                     COIN_R, MUSHROOM_R)
from .palette import INK, CREAM, SKY, DKRED, ORNG, GOLD, RED, WHITE
from .sprites import (draw_parrot_tilt, draw_trail, draw_pipe,
                      draw_coin, draw_mushroom)


class Bird:
    TRAIL_LEN = 10

    def __init__(self):
        self.reset()

    def reset(self):
        self.x   = float(BIRD_X)
        self.y   = 120.0
        self.vy  = 0.0
        self.alive = True
        self.t   = 0.0
        self.trail: list[tuple[float, float]] = []
        self._flap_t = 0.0   # time since last flap (drives wing cycle)

    def flap(self):
        if not self.alive:
            return
        self.vy = FLAP_V
        self._flap_t = 0.0   # restart wing cycle from top

    def update(self, dt):
        self.vy = min(self.vy + GRAVITY * dt, MAX_FALL)
        self.y += self.vy * dt
        self.t += dt
        self._flap_t += dt
        self.trail.append((self.x, self.y))
        if len(self.trail) > self.TRAIL_LEN:
            self.trail.pop(0)

    @property
    def frame(self):
        # Wing sweeps down over ~0.35 s then resets
        cycle = 0.35
        phase = (self._flap_t % cycle) / cycle       # 0→1
        if self.vy > 180:
            return 3                                  # diving: wings down
        return int(phase * 4) % 4

    def hits_rect(self, rx, ry, rw, rh):
        """Circle-vs-AABB test."""
        cx = max(rx, min(self.x, rx + rw))
        cy = max(ry, min(self.y, ry + rh))
        dx, dy = self.x - cx, self.y - cy
        return dx*dx + dy*dy <= BIRD_R * BIRD_R

    def draw(self):
        draw_trail(self.trail)
        draw_parrot_tilt(self.x, self.y, self.frame, self.vy)


# ─────────────────────────────────────────────────────────────────────────────

class Pipe:
    def __init__(self, x, gap_y, gap_h):
        self.x = float(x)
        self.gap_y = gap_y
        self.gap_h = gap_h
        self.passed = False

    def update(self, dt, speed):
        self.x -= speed * dt

    def draw(self):
        draw_pipe(int(self.x), self.gap_y, self.gap_h, GROUND_Y)

    @property
    def top_rect(self):
        return (self.x, 0, PIPE_W, self.gap_y)

    @property
    def bot_rect(self):
        by = self.gap_y + self.gap_h
        return (self.x, by, PIPE_W, max(0, GROUND_Y - by))


# ─────────────────────────────────────────────────────────────────────────────

class Coin:
    def __init__(self, x, y):
        self.x = float(x)
        self.y = float(y)
        self.alive = True
        self.t = random.random() * 10

    def update(self, dt, speed):
        self.x -= speed * dt
        self.t += dt

    def draw(self):
        draw_coin(self.x, self.y, self.t)

    def hits(self, bird):
        dx = self.x - bird.x
        dy = (self.y + math.sin(self.t * 4.5) * 1.5) - bird.y
        r  = COIN_R + BIRD_R
        return dx*dx + dy*dy <= r*r


# ─────────────────────────────────────────────────────────────────────────────

class Mushroom:
    def __init__(self, x, y):
        self.x = float(x)
        self.y = float(y)
        self.alive = True
        self.t = 0.0

    def update(self, dt, speed):
        self.x -= speed * dt
        self.t += dt

    def draw(self):
        draw_mushroom(self.x, self.y, self.t)

    def hits(self, bird):
        dx = self.x - bird.x
        dy = self.y - bird.y
        r  = MUSHROOM_R + BIRD_R
        return dx*dx + dy*dy <= r*r


# ─────────────────────────────────────────────────────────────────────────────

class Particle:
    def __init__(self, x, y, vx, vy, life, col, size=1, grav=0.0):
        self.x, self.y = float(x), float(y)
        self.vx, self.vy = float(vx), float(vy)
        self.life = self.max_life = float(life)
        self.col  = col
        self.size = size
        self.grav = grav
        self.alive = True

    def update(self, dt):
        self.life -= dt
        if self.life <= 0:
            self.alive = False
            return
        self.vy += self.grav * dt
        self.x  += self.vx * dt
        self.y  += self.vy * dt

    def draw(self):
        if self.size <= 1:
            _px = pyxel.pset
            _px(int(self.x), int(self.y), self.col)
        else:
            s = self.size
            pyxel.rect(int(self.x) - s//2, int(self.y) - s//2, s, s, self.col)


class FloatText:
    def __init__(self, x, y, text, col):
        self.x, self.y = float(x), float(y)
        self.text = text
        self.col  = col
        self.life = 0.75
        self.alive = True

    def update(self, dt):
        self.life -= dt
        self.y    -= 25 * dt
        if self.life <= 0:
            self.alive = False

File: src/entities.js
import {
    GRAVITY, FLAP_V, MAX_FALL, BIRD_X, BIRD_R, GROUND_Y,
    PIPE_W, COIN_R,
} from './config.js';
import { C } from './palette.js';
import { BIRD_FRAMES, COIN_FRAMES, MUSHROOM, PIPE_BODY, PIPE_CAP_28 } from './sprites.js';

export class Bird {
    constructor() {
        this.x = BIRD_X;
        this.y = 140;
        this.vy = 0;
        this.r = BIRD_R;
        this.alive = true;
        this.animT = 0;
        this.trail = [];
        this.flapPulse = 0; // non-zero briefly after a flap (for bigger wing)
    }

    reset() {
        this.x = BIRD_X; this.y = 140; this.vy = 0;
        this.alive = true; this.animT = 0; this.trail = []; this.flapPulse = 0;
    }

    flap() {
        if (!this.alive) return;
        this.vy = FLAP_V;
        this.flapPulse = 0.12;
    }

    update(dt) {
        this.vy += GRAVITY * dt;
        if (this.vy > MAX_FALL) this.vy = MAX_FALL;
        this.y += this.vy * dt;
        this.animT += dt;
        if (this.flapPulse > 0) this.flapPulse -= dt;

        // Trail
        this.trail.push({ x: this.x, y: this.y });
        if (this.trail.length > 8) this.trail.shift();
    }

    frame() {
        // Fast flutter; force wing-down when falling fast.
        if (this.vy > 180) return 2;
        const t = this.animT;
        const f = Math.floor(t * 14) % 3;
        return f;
    }

    angle() {
        // Nose tilt based on vy: up when flapping, down when diving.
        const a = this.vy / 380;
        return Math.max(-0.5, Math.min(1.0, a));
    }

    draw(buf) {
        // Draw trail first (4-point low-alpha ghost circles).
        for (let i = 0; i < this.trail.length; i++) {
            const p = this.trail[i];
            const k = i / this.trail.length;
            if (k < 0.4) continue;
            const r = 2 + k * 2;
            buf.disc(p.x, p.y, r, k > 0.8 ? C.CREAM : C.DUSK);
        }
        const spr = BIRD_FRAMES[this.frame()];
        buf.blit(spr, Math.round(this.x - spr.w / 2), Math.round(this.y - spr.h / 2));
    }

    // Bounding circle hit test used against pipes / ground.
    hitsRect(rx, ry, rw, rh) {
        const cx = Math.max(rx, Math.min(this.x, rx + rw));
        const cy = Math.max(ry, Math.min(this.y, ry + rh));
        const dx = this.x - cx, dy = this.y - cy;
        return (dx * dx + dy * dy) <= this.r * this.r;
    }
}

export class Pipe {
    constructor(x, gapY, gapH) {
        this.x = x;                // left edge
        this.gapY = gapY;          // top of the gap
        this.gapH = gapH;
        this.passed = false;
        this.spawnIndex = 0;       // used to avoid double-spawning collectibles
        this.w = PIPE_W;
    }

    get topRect() { return { x: this.x, y: 0, w: this.w, h: this.gapY }; }
    get botRect() { return { x: this.x, y: this.gapY + this.gapH, w: this.w, h: GROUND_Y - (this.gapY + this.gapH) }; }

    update(dt, scrollSpeed) {
        this.x -= scrollSpeed * dt;
    }

    draw(buf) {
        const topH = this.gapY;
        const botY = this.gapY + this.gapH;
        const botH = GROUND_Y - botY;

        // Top pipe body (descending) then cap at the bottom of this segment.
        const bodyTopH = Math.max(0, topH - PIPE_CAP_28.h);
        for (let y = 0; y < bodyTopH; y++) {
            // repeat the 1-row body
            for (let i = 0; i < PIPE_BODY.w; i++) {
                const ci = PIPE_BODY.data[i];
                if (ci < 0) continue;
                buf.put(this.x + i, y, ci);
            }
        }
        if (topH >= PIPE_CAP_28.h) {
            buf.blit(PIPE_CAP_28, this.x, bodyTopH);
        }

        // Bottom pipe: cap first, body below.
        if (botH >= PIPE_CAP_28.h) {
            buf.blit(PIPE_CAP_28, this.x, botY);
            for (let y = botY + PIPE_CAP_28.h; y < GROUND_Y; y++) {
                for (let i = 0; i < PIPE_BODY.w; i++) {
                    const ci = PIPE_BODY.data[i];
                    if (ci < 0) continue;
                    buf.put(this.x + i, y, ci);
                }
            }
        }
    }
}

export class Coin {
    constructor(x, y) {
        this.x = x; this.y = y;
        this.r = COIN_R;
        this.alive = true;
        this.t = Math.random() * 10;
        this.bobY = 0;
    }

    update(dt, scrollSpeed) {
        this.x -= scrollSpeed * dt;
        this.t += dt;
        this.bobY = Math.sin(this.t * 4) * 1.2;
    }

    draw(buf) {
        const frame = Math.floor(this.t * 8) % COIN_FRAMES.length;
        const spr = COIN_FRAMES[frame];
        buf.blit(spr, Math.round(this.x - spr.w / 2), Math.round(this.y + this.bobY - spr.h / 2));
    }

    hits(bird) {
        const dx = this.x - bird.x;
        const dy = this.y + this.bobY - bird.y;
        const rr = this.r + bird.r;
        return dx * dx + dy * dy <= rr * rr;
    }
}

export class Mushroom {
    constructor(x, y) {
        this.x = x; this.y = y;
        this.r = 6;
        this.alive = true;
        this.t = 0;
    }

    update(dt, scrollSpeed) {
        this.x -= scrollSpeed * dt;
        this.t += dt;
    }

    draw(buf) {
        // Gentle bob + a faint halo ring that pulses.
        const bob = Math.sin(this.t * 3) * 1.5;
        const r = 9 + Math.sin(this.t * 4) * 1.2;
        buf.ring(this.x, this.y + bob, r, C.CREAM);
        buf.ring(this.x, this.y + bob, r + 2, C.ORANGE);
        buf.blit(MUSHROOM, Math.round(this.x - MUSHROOM.w / 2), Math.round(this.y + bob - MUSHROOM.h / 2));
    }

    hits(bird) {
        const dx = this.x - bird.x;
        const dy = this.y - bird.y;
        const rr = this.r + bird.r;
        return dx * dx + dy * dy <= rr * rr;
    }
}

export class Particle {
    constructor(x, y, vx, vy, life, color, size = 1, gravity = 0) {
        this.x = x; this.y = y;
        this.vx = vx; this.vy = vy;
        this.life = life; this.maxLife = life;
        this.color = color;
        this.size = size;
        this.gravity = gravity;
        this.alive = true;
    }

    update(dt) {
        this.life -= dt;
        if (this.life <= 0) { this.alive = false; return; }
        this.vy += this.gravity * dt;
        this.x += this.vx * dt;
        this.y += this.vy * dt;
    }

    draw(buf) {
        if (this.size <= 1) buf.put(this.x, this.y, this.color);
        else buf.rect(this.x - (this.size / 2 | 0), this.y - (this.size / 2 | 0), this.size, this.size, this.color);
    }
}

// Floating "+1" / "+3" labels.
export class FloatText {
    constructor(x, y, text, color) {
        this.x = x; this.y = y;
        this.text = text;
        this.color = color;
        this.life = 0.7;
        this.maxLife = 0.7;
        this.alive = true;
    }

    update(dt) {
        this.life -= dt;
        this.y -= 22 * dt;
        if (this.life <= 0) this.alive = false;
    }
}
